### PR TITLE
Add an indicator for AES GCM that returns if the iv has been generated

### PR DESCRIPTION
**Pull request by slontis**
internally.

This is not using a strict check since there may be applications that require the IV to be generated externally (e.g. java).

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/acvp_test.c [hunk 1]
```diff
@@ -853,34 +854,53 @@ static int aes_gcm_enc_dec(const char *alg,
...
+        unsigned int iv_generated = -1;
+
+        params[0] = OSSL_PARAM_construct_uint(OSSL_CIPHER_PARAM_AEAD_IV_GENERATED,
+                                              &iv_generated);
+        if (outiv != NULL)
+            params[1] = OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_IV,
+                                                          outiv, iv_len);
+        if (!TEST_true(EVP_CIPHER_CTX_get_params(ctx, params)))
+                goto err;
+        printf("%d %d %d\n", iv_generated, enc, iv != NULL);
```
Comment (Sashan) [inline test/acvp_test.c:897]: Are we sure we want to keep the `printf()` line as-is? It looks like a kind of debugging aid now.

**Review by Sashan — CHANGES_REQUESTED** [review]
changes look good to me. I think just one nit in unit test.

File: test/acvp_test.c (hunk 1 above)
Comment (slontis) [reply test/acvp_test.c:897]: no - that was me debugging :)

File: doc/man3/EVP_EncryptInit.pod [hunk 2]
```diff
@@ -905,6 +905,16 @@ A getter that returns 1 if the operation is FIPS approved, or 0 otherwise.
 This may be used after calling a cipher final operation such as
 EVP_EncryptFinal_ex(). It may return 0 if the "encrypt-check" option is set to 0.
 
+=item "iv-generated" (B<OSSL_CIPHER_PARAM_AEAD_IV_GENERATED>) <unsigned integer>
+
+An indicator that returns 1 if an IV was generated internally during encryption,
+or O otherwise.
+This may be used by AES GCM ciphers after calling a cipher final operation such
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:912]: Maybe drop AES?  This applies to all GCM mode ciphers.

File: doc/man3/EVP_EncryptInit.pod [hunk 3]
```diff
@@ -905,6 +905,16 @@ A getter that returns 1 if the operation is FIPS approved, or 0 otherwise.
...
+as EVP_EncryptFinal_ex().
+AES GCM should generate an IV internally if the iv is not specified during a
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:914]: Keep case consistent, it's `IV` not `iv`.

Also drop AES here if dropped above.

File: doc/man3/EVP_EncryptInit.pod [hunk 4]
```diff
@@ -905,6 +905,16 @@ A getter that returns 1 if the operation is FIPS approved, or 0 otherwise.
...
+cipher init call such as EVP_CipherInit_ex().
```
Comment (paulidale) [inline doc/man3/EVP_EncryptInit.pod:915]: s/init/initialisation/

File: test/acvp_test.c [hunk 5]
```diff
@@ -853,34 +854,53 @@ static int aes_gcm_enc_dec(const char *alg,
...
     if (enc) {
-        if (!TEST_mem_eq(out, out_len, ct, ct_len)
-            || !TEST_int_gt(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG,
-                                              tag_len, out + out_len), 0)
-            || !TEST_mem_eq(out + out_len, tag_len, tag, tag_len))
+        if ((ct != NULL && !TEST_mem_eq(out, olen, ct, ct_len))
+                || !TEST_int_gt(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG,
+                                                    tag_len, out + olen), 0)
+                || (tag != NULL
+                   && !TEST_mem_eq(out + olen, tag_len, tag, tag_len)))
```
Comment (paulidale) [inline test/acvp_test.c:879]: indentation nit

**Review by paulidale — APPROVED** [review]
approved with nits fixed

File: doc/man3/EVP_EncryptInit.pod (hunk 3 above)
Comment (paulidale) [reply doc/man3/EVP_EncryptInit.pod]: You changed them all to lower case which is wrong.  It should be IV.
There were both in this line, hence the confusion.

**Review by paulidale — CHANGES_REQUESTED** [review]
Should be IV (capital)

**Review by Sashan — APPROVED** [review]
looks good to me as far as I can tell.
thanks.

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
@paulidale ?

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**